### PR TITLE
[#7247] Bumped Boost version to 1.81 (main)

### PR DESCRIPTION
**Pull request by korydraughn**
This has not been tested.

Requires irods/externals#194.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,8 +22,8 @@ macro(IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_DEVELOP_DEPENDENCIE
 endmacro()
 
 IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIES_LIST(ARCHIVE libarchive3.5.2-0)
-IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIES_LIST(AVRO avro1.11.0-1)
-IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIES_LIST(BOOST boost1.78.0-0)
+IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIES_LIST(AVRO avro1.11.0-0)
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply CMakeLists.txt:25]: Right ...

Seems we need to consider what the rules are for handling the consortium build number.

@SwooshyCueb Thoughts?

[the remark this answers is not in the document]
Comment (trel) [reply CMakeLists.txt:25]: consortium build number gets bumped on the dependency of whatever library got version bumped.

so if a thing is built using boost, and its own version number is not getting bumped, its consortium build number would need to go from 0 to 1.

[the remark this answers is not in the document]
Comment (alanking) [reply CMakeLists.txt:25]: I made the comment before I actually looked at irods/externals#194. Wasn't sure why it went backwards but now I see. I think we need to either change the package name or bump the build number to 2, per the conversation in that PR. We've already released build number 0, I believe.

[the remark this answers is not in the document]
Comment (korydraughn) [reply CMakeLists.txt:25]: My bad, I misread the comment at https://github.com/irods/irods/issues/7247#issuecomment-1664359518.

It needs to be bumped.

Will change.

File: CMakeLists.txt [hunk 2]
```diff
@@ -240,7 +240,7 @@ set(
 # latest version of a compiler or platform.
 # See the following for more information:
 #   https://www.boost.org/doc/libs/1_77_0/doc/html/boost_asio/using.html
-list(APPEND IRODS_COMPILE_DEFINITIONS_PRIVATE "BOOST_ASIO_HAS_STD_INVOKE_RESULT")
+#list(APPEND IRODS_COMPILE_DEFINITIONS_PRIVATE "BOOST_ASIO_HAS_STD_INVOKE_RESULT") TODO Perhaps this isn't necessary for 1.81.
```
Comment (korydraughn) [inline CMakeLists.txt:243]: The server now compiles without this.

If all unit tests and core tests pass, then it is safe to remove.

**Comment by korydraughn** [comment]
All unit tests passed.

Running core tests.

**Review by SwooshyCueb — APPROVED** [review]

**Comment by korydraughn** [comment]
All core tests passed.

File: CMakeLists.txt (hunk 2 above)
Comment (korydraughn) [reply CMakeLists.txt]: Removed

File: CMakeLists.txt [hunk 3]
```diff
@@ -234,13 +234,6 @@ set(
   SPDLOG_FMT_EXTERNAL
   SPDLOG_NO_TLS
 )
-# Manually defined because Boost cannot detect that std::result_of has been removed in C++20.
-# Verified that Boost 1.77 and Boost 1.78 do not handle this yet. Normally, defining
-# BOOST_ASIO_HAS_ macros should be avoided except when absolutely required as a workaround for the
-# latest version of a compiler or platform.
-# See the following for more information:
-#   https://www.boost.org/doc/libs/1_77_0/doc/html/boost_asio/using.html
-list(APPEND IRODS_COMPILE_DEFINITIONS_PRIVATE "BOOST_ASIO_HAS_STD_INVOKE_RESULT")
```
Comment (trel) [inline CMakeLists.txt:243]: is anything else using this?  needs to be updated?

Comment (korydraughn) [reply CMakeLists.txt:243]: It is inherited by other projects because they use `IRODS_COMPILE_DEFINITIONS_PRIVATE`.

Comment (SwooshyCueb) [inline CMakeLists.txt:243]: Given that many of the distros we support have boost 1.7X in their repos, let's document somewhere that we needed this for boost 1.77-1.80. This information may be useful once we start relying on distro-provided boost.

File: lib/core/src/irods_path_recursion.cpp [hunk 4]
```diff
@@ -226,7 +226,8 @@ irods::check_directories_for_loops( boost::filesystem::path const & dirpath,
     try {
         // Default constructor creates an end iterator
         fs::recursive_directory_iterator end_itr;
-        const fs::symlink_option opt = (dashdashlink? fs::symlink_option::none : fs::symlink_option::recurse);
+        const fs::directory_options opt =
+            (dashdashlink ? fs::directory_options::none : fs::directory_options::follow_directory_symlink);
```
Comment (SwooshyCueb) [inline lib/core/src/irods_path_recursion.cpp:230]: Given that many of the distros we support have boost 1.7X in their repos, let's document somewhere that we made this change for boost 1.81 compatibility when upgrading from 1.78. This information may be useful once we start relying on distro-provided boost.

**Review by SwooshyCueb — APPROVED** [review]
LGTM, just a couple of things I want written down somewhere. (I might do it myself if nobody else gets to it before this PR is cherry-picked.)

File: CMakeLists.txt (hunk 3 above)
Comment (korydraughn) [reply CMakeLists.txt:243]: Good idea.

File: lib/core/src/irods_path_recursion.cpp (hunk 4 above)
Comment (korydraughn) [reply lib/core/src/irods_path_recursion.cpp:230]: Good idea.

File: CMakeLists.txt (hunk 3 above)
Comment (korydraughn) [reply CMakeLists.txt:243]: Capture in issue #7282.

File: lib/core/src/irods_path_recursion.cpp (hunk 4 above)
Comment (korydraughn) [reply lib/core/src/irods_path_recursion.cpp:230]: Capture in issue #7282.

**Review by alanking — APPROVED** [review]
Please # once builds come back green

**Comment by korydraughn** [comment]
All github actions succeeded.

#'d

**Review by alanking — APPROVED** [review]

**Pull request merged**